File: backend/app/schemas/billing.py
from pydantic import BaseModel
from typing import List, Optional


class UsageItem(BaseModel):
    label: str
    used: int
    limit: str
    pct: int
    remaining: Optional[int] = None


class CurrentPlan(BaseModel):
    plan: str  # free, starter, pro
    price: float
    billing_cycle: str = "monthly"
    next_billing_date: Optional[str] = None
    features: List[str]
    usage: List[UsageItem]
    usage_reset_date: Optional[str] = None


class CheckoutRequest(BaseModel):
    plan: str  # starter, pro
    billing_cycle: str = "monthly"  # monthly, yearly


class CheckoutResponse(BaseModel):
    checkout_url: str
    session_id: str


class Invoice(BaseModel):
    id: str
    date: str
    amount: float
    currency: str
    status: str
    pdf_url: Optional[str] = None


class PortalResponse(BaseModel):
    portal_url: str

File: frontend/lib/api/billing.ts
import { request } from "./client";

export interface UsageItem {
  label: string;
  used: number;
  limit: string;
  pct: number;
  remaining?: number;
}

export interface CurrentPlan {
  plan: "free" | "starter" | "pro";
  price: number;
  billing_cycle: string;
  next_billing_date?: string;
  features: string[];
  usage: UsageItem[];
  usage_reset_date?: string;
}

export interface Invoice {
  id: string;
  date: string;
  amount: number;
  currency: string;
  status: string;
  pdf_url?: string;
}

export const billingApi = {
  getPlan: () => request<CurrentPlan>("/billing/plan"),

  createCheckout: (plan: "starter" | "pro", billingCycle: "monthly" | "yearly" = "monthly") =>
    request<{ checkout_url: string; session_id: string }>("/billing/checkout", {
      method: "POST",
      body: JSON.stringify({ plan, billing_cycle: billingCycle }),
    }),

  getHistory: () => request<Invoice[]>("/billing/history"),

  createPortal: () =>
    request<{ portal_url: string }>("/billing/portal", { method: "POST" }),
};
